src/libmeasurement_kit/dns/getaddrinfo_async.hpp

```diff
@@ -0,0 +1,121 @@
+// Part of measurement-kit <https://measurement-kit.github.io/>.
+// Measurement-kit is free software. See AUTHORS and LICENSE for more
+// information on the copying conditions.
+
+#include <measurement_kit/dns.hpp>
+
+namespace mk {
+namespace dns {
+
+#define GETADDRINFO_ASYNC_MAP_ERROR                                            \
+    XX(0, NoError)                                                             \
+    XX(EAI_AGAIN, TemporaryFailureError)                                       \
+    XX(EAI_BADFLAGS, InvalidFlagsValueError)                                   \
+    XX(EAI_FAIL, NonRecoverableFailureError)                                   \
+    XX(EAI_FAMILY, NotSupportedAIFamilyError)                                  \
+    XX(EAI_MEMORY, MemoryAllocationFailureError)                               \
+    XX(EAI_NONAME, HostOrServiceNotProvidedOrNotKnownError)                    \
+    XX(EAI_OVERFLOW, ArgumentBufferOverflowError)                              \
+    XX(EAI_SERVICE, NotSupportedServnameError)                                 \
+    XX(EAI_SOCKTYPE, NotSupportedAISocktypeError)
```

Comment:
I would suggest we leave these as separate errors (in terms of `Error` definition), but map the string encoding of the error to a set of known ooni error code.

The mapping I would suggest is as follows (for reference ooni error string [are defined here](https://github.com/TheTorProject/ooni-spec/blob/master/data-formats/df-000-base.md#error-strings))

* `EAI_AGAIN`->`dns_name_failure` (Note: This could create confusion in ooniprobe as twisted does not map this `EAGAIN`)
* `EAI_BADFLAGS `->`unknown_failure dns_bad_flags`
* `EAI_FAIL, NonRecoverableFailureError `->`dns_name_failure`
* `EAI_FAMILY, NotSupportedAIFamilyError `->`unknown_failure dns_family_failure`
* `EAI_MEMORY, MemoryAllocationFailureError `->`unknown_failure dns_memory_failure`
* `EAI_NONAME, HostOrServiceNotProvidedOrNotKnownError `->`dns_name_error`
* `EAI_OVERFLOW, ArgumentBufferOverflowError `->`unknown_failure dns_overflow_failure`
* `EAI_SERVICE, NotSupportedServnameError `->`unknown_failure dns_service_failure` (I think the mapping in MK `NotSupportedServnameError` is misleading. I would instead write it as `NotSupportedServiceError` as `Servname` seems to imply it has something to do with DNS server)
* `EAI_SOCKTYPE, NotSupportedAISocktypeError `->`unknown_failure dns_socktype_failure`